pkg/health/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,277 @@
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/coordination/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/config"
+	fake "sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+// Heartbeat is a struct for the heartbeat package
+type Heartbeat struct {
+	logger         *logrus.Entry
+	heartbeatTimer *time.Timer
+	etcdConfig     *etcdutil.EtcdConnectionConfig
+	leaseDuration  int32
+	K8sClient      client.Client
+}
+
+// NewHeartbeat returns the heartbeat object.
+func NewHeartbeat(hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *etcdutil.EtcdConnectionConfig) (*Heartbeat, error) {
+	leaseDuration := int32(hconfig.LeaseDuration.Seconds())
+	if leaseDuration < 1 {
+		return nil, &errors.EtcdError{
+			Message: "Could not initialize new heartbeat: Duration provided is too small",
+		}
+	}
+	restConfig, err := config.GetConfig()
+	var cl client.Client
+	if err == nil {
+		cl, err = client.New(restConfig, client.Options{})
+	}
+	if err != nil {
+		logger.Info("could not load InCluster config")
+		cl = fake.NewClientBuilder().Build()
+	}
+	return &Heartbeat{
+		logger:        logger.WithField("actor", "heartbeat"),
+		etcdConfig:    etcdConfig,
+		leaseDuration: leaseDuration,
+		K8sClient:     cl,
+	}, nil
+}
+
+// RenewMemberLeases is called periodically and renews the pods member lease
+func (hb *Heartbeat) RenewMemberLeases() error {
+	//Fetch lease associated with member
+	memberName, err := GetEnvVarOrError("POD_NAME")
+	fmt.Println("member name is ", memberName)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("POD NAME env var not present: %v", err),
+		}
+	}
+	namespace, err := GetEnvVarOrError("POD_NAMESPACE")
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("POD_NAMESPACE env var not present: %v", err),
+		}
+	}
+	memberLease := &v1.Lease{}
+	err = hb.K8sClient.Get(context.TODO(), client.ObjectKey{
+		Namespace: namespace,
+		Name:      memberName,
+	}, memberLease)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not fetch member lease: %v", err),
+		}
+	}
+
+	//Create etcd client to get etcd ID
+	etcdClient, err := etcdutil.GetTLSClientForEtcd(hb.etcdConfig)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to create etcd client: %v", err),
+		}
+	}
+	defer etcdClient.Close()
+	response, err := etcdClient.Status(context.TODO(), hb.etcdConfig.Endpoints[0])
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Failed to get status of etcd endPoint [ %v ] with error: %v", hb.etcdConfig.Endpoints[0], err),
+		}
+	}
+	memberID := ""
+	if response.Header.MemberId == response.Leader {
+		memberID = strconv.FormatUint(response.Header.MemberId, 16) + ":Leader"
+	} else {
+		memberID = strconv.FormatUint(response.Header.MemberId, 16) + ":Member"
```

Comment:
Since the `strconv.FormatUint(response.Header.MemberId, 16)` is common to both paths in the `if` construct, you can move it out of the `if` block to avoid redundancy in computation.

Reply:
Updated this